packages/core/content-releases/server/src/services/release-action.ts

```diff
@@ -0,0 +1,340 @@
+        validateEntryData(action.contentType, action.entryDocumentId),
+        validateUniqueEntry(releaseId, action),
+      ]);
+
+      const release = await strapi.db
+        .query(RELEASE_MODEL_UID)
+        .findOne({ where: { id: releaseId } });
+
+      if (!release) {
+        throw new errors.NotFoundError(`No release found for id ${releaseId}`);
+      }
+
+      if (release.releasedAt) {
+        throw new errors.ValidationError('Release already published');
+      }
+
+      // If the action is a publish, check if the entry is valid
+      // If the action is an unpublish, skip the validation
+      const actionStatus =
+        action.type === 'publish'
+          ? getDraftEntryValidStatus(
+              {
+                contentType: action.contentType,
+                documentId: action.entryDocumentId,
+                locale: action.locale,
+              },
+              {
+                strapi,
+              }
+            )
+          : true;
+
+      const releaseAction = await strapi.db.query(RELEASE_ACTION_MODEL_UID).create({
+        data: {
+          ...action,
+          release: release.id,
+          isEntryValid: actionStatus,
+        },
+        populate: { release: { select: ['id'] } },
+      });
+
+      getService('release', { strapi }).updateReleaseStatus(release.id);
+
+      return releaseAction;
+    },
+
+    async findPage(
+      releaseId: GetReleaseActions.Request['params']['releaseId'],
+      query?: GetReleaseActions.Request['query']
+    ) {
+      const release = await strapi.db.query(RELEASE_MODEL_UID).findOne({
+        where: { id: releaseId },
+        select: ['id'],
+      });
+
+      if (!release) {
+        throw new errors.NotFoundError(`No release found for id ${releaseId}`);
+      }
+
+      const dbQuery = strapi.get('query-params').transform(RELEASE_ACTION_MODEL_UID, query ?? {});
+      const { results: actions, pagination } = await strapi.db
+        .query(RELEASE_ACTION_MODEL_UID)
+        .findPage({
+          ...dbQuery,
+          where: {
+            release: releaseId,
+          },
+        });
+
+      // For each contentType on the release, we create a custom populate object for nested relations
+      const populateBuilderService = strapi.plugin('content-manager').service('populate-builder');
+
+      const actionsWithEntry = await async.map(actions, async (action: ReleaseAction) => {
+        // @ts-expect-error - Core.Service type is not a function
+        const populate = await populateBuilderService(action.contentType)
+          .populateDeep(Infinity)
+          .build();
+
+        return {
+          ...action,
+          entry: await getEntry(
+            {
+              contentType: action.contentType,
+              documentId: action.entryDocumentId,
+              locale: action.locale,
+              populate,
+              status: action.type === 'publish' ? 'draft' : 'published',
+            },
+            { strapi }
+          ),
```

Comment:
what happens if the content type doesnt exist? (it has been deleted)

Reply:
That case should be handle by the functions on db lifecycles, currently we delete actions if content type is deleted or D&P is disabled on the Content Type